datacollectiondemo/source/cpp/src/KaaDemo.cpp

```diff
@@ -14,77 +14,110 @@
-     * Run the Kaa endpoint.
-     */
-    kaaClient->start();
+    void run()
+    {
+        // Run the Kaa endpoint.
+        kaaClient_->start();
+        samplePeriod_ = kaaClient_->getConfiguration().samplePeriod;
+        timer_.async_wait(boost::bind(&TemperatureSensor::sendTemperature, this));
+        timer_.expires_from_now(boost::posix_time::seconds(samplePeriod_));
+        service_.run();
+    }
 
-    std::list<std::pair<RecordFuture, std::size_t>> futurePairs;
+private:
+    static constexpr auto savedConfig_ = "saved_config.cfg";
+    std::shared_ptr<IKaaClient> kaaClient_;
+    int32_t samplePeriod_;
+    boost::asio::io_service service_;
+    boost::posix_time::seconds interval_;
+    boost::asio::deadline_timer timer_;
+    std::mutex iostream_mutex_;
+
+    int32_t getTemperature(void)
+    {
+        // For sake of example random data is used
+        return rand() % 10 + 25;
+    }
 
-    // Send LOGS_TO_SEND_COUNT logs in a loop.
-    std::size_t logNumber = 0;
-    while (logNumber++ < LOGS_TO_SEND_COUNT) {
+    void sendTemperature()
+    {
         KaaUserLogRecord logRecord;
-        logRecord.level = kaa_log::Level::KAA_INFO;
-        logRecord.tag = "TAG";
-        logRecord.message = "MESSAGE_" + std::to_string(logNumber);
-        logRecord.timeStamp = TimeUtils::getCurrentTimeInMs();
-
-        futurePairs.push_back(std::make_pair(std::move(kaaClient->addLogRecord(logRecord)), logRecord.timeStamp));
-        std::cout << "Sent " << logNumber << "th record" << std::endl;
+        logRecord.temperature = getTemperature();
+        logRecord.timeStamp = std::time(nullptr);
+        // Send value of temperature
+        kaaClient_->addLogRecord(logRecord);
+        // Show log
+        iostream_mutex_.lock();
+        std::cout << "Sampled temperature: " << logRecord.temperature <<
+            ", timestamp: " << logRecord.timeStamp << std::endl;
+        iostream_mutex_.unlock();
+        // Change timer expiry period
+        timer_.expires_from_now(boost::posix_time::seconds(samplePeriod_));
+        // Posts the timer event
+        timer_.async_wait(boost::bind(&TemperatureSensor::sendTemperature, this));
     }
 
-    for (auto& pair : futurePairs) {
-        try {
-            RecordInfo recordInfo = pair.first.get();
-            BucketInfo bucketInfo = recordInfo.getBucketInfo();
-
-            std::size_t timeSpent = (recordInfo.getRecordAddedTimestampMs() - pair.second)
-                                        + recordInfo.getRecordDeliveryTimeMs();
-
-            std::cout << "Received log record delivery info. Bucket Id [" <<  bucketInfo.getBucketId() << "]. "
-                      << "Record delivery time [" << timeSpent << " ms]." << std::endl;
-        } catch (std::exception& e) {
-            std::cout << "Exception was caught while waiting for callback future" << e.what();
-        }
+    void onConfigurationUpdated(const KaaRootConfiguration &configuration)
+    {
+        iostream_mutex_.lock();
```

Comment:
_Resource leak_
[operator <<] (http://ru.cppreference.com/w/cpp/io/basic_ostream/operator_ltlt) lacks `noexcept` or `throw()` identifier. You are to use guard.